Google.Api.Gax.Grpc/GrpcCoreAdapter.cs

```diff
@@ -0,0 +1,165 @@
+    /// </summary>
+    public sealed class GrpcCoreAdapter : GrpcAdapter
+    {
+        private const string GrpcCoreAssemblyName = "Grpc.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=d754f35622e28bad";
+
+        // Option names, copied from ChannelOptions
+        internal const string ServiceConfigDisableResolution = "grpc.service_config_disable_resolution";
+        internal const string KeepAliveTimeMs = "grpc.keepalive_time_ms";
+        internal const string KeepAliveTimeoutMs = "grpc.keepalive_timeout_ms";
+        internal const string MaxReceiveMessageLength = "grpc.max_receive_message_length";
+        internal const string MaxSendMessageLength = "grpc.max_send_message_length";
+        internal const string PrimaryUserAgentString = "grpc.primary_user_agent";
+
+        // Note: we create the channel factory lazily rather than as part of type initialization as it can
+        // throw if the dependency is missing. Type initialization failing can make all kinds of things messy.
+        private static Lazy<Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase>> channelFactory =
+            new Lazy<Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase>>(CreateFactory, LazyThreadSafetyMode.PublicationOnly);
+
+        /// <summary>
+        /// Prevent lazy type initialization.
+        /// </summary>
+        static GrpcCoreAdapter()
+        {
+        }
+
+        /// <summary>
+        /// Returns the singleton instance of this class.
+        /// </summary>
+        public static GrpcCoreAdapter Instance { get; } = new GrpcCoreAdapter();
+
+        private GrpcCoreAdapter()
+        {
+        }
+
+        /// <inheritdoc />
+        protected override ChannelBase CreateChannelImpl(string endpoint, ChannelCredentials credentials, GrpcChannelOptions options) =>
+            channelFactory.Value.Invoke(endpoint, credentials, options);
+
+        private static Func<string, ChannelCredentials, GrpcChannelOptions, ChannelBase> CreateFactory()
+        {
+            var optionType = System.Type.GetType("Grpc.Core.ChannelOption, " + GrpcCoreAssemblyName);
+            if (optionType is null)
+            {
+                throw new InvalidOperationException("Unable to find Grpc.Core types. Please ensure you have a suitable Grpc.Core dependency");
+            }
+
+            // Result should be equivalent to
+            // new Channel(endpoint, credentials, ConvertOptions(options));
+
+            // TODO: Benchmark using expressions vs invoking the Channel constructor by reflection.
+            // The latter may be more friendly to environments where expression trees do not work -
+            // but in those environments, it's unlikely that Grpc.Core is being used anyway.
+            // The other methods can be invoked with delegates, but delegates can't refer directly to methods.
+
+            // TODO: Somehow cache the result of ConvertOptions? It's probably not called terribly frequently.
```

Comment:
- Is ChannelOption inmutable?
- Consider caching each individual option instead of the list? And you can pick and choose based on the value?

Reply:
Yes, both ChannelOption and GrpcChanneloptions are immutable. Caching individual options would be trickier, given that we still need to come up with a sequence of the results. I suspect it's probably simpler to do all or nothing (and let's do nothing until we know we need to do anything...)